### PR TITLE
Allow to specify a default for bcond features in a macro file

### PR DESCRIPTION
**Pull request by mlschroe**
The bcond mechanism of rpm allows to specify a default for a feature in a spec file, which can be overridden with the command line. SUSE wants to build packages with the same source in different projects. Thus we need a way to specify a default via an rpm macro.

Our first try was to misuse the command line feature by putting `%_with_foo 1` in the macros. But that has two disadvantages: it breaks the command line and it makes it impossible to change the default in a derived project, as rpm cannot undefine macros in a macrofile and the bcond implementation only checks for definedness.

So here's a small addition that makes it possible to set feature defaults: Feature `foo` can be enabled by adding
the macro
```
%bcond_default_foo 1
```
The mechanism uses expression evaluation, so it is possible to remove a default by using %nil as default.

Maybe we can use a better name as `bcond_default`. Any suggestions?

(The pull request is missing documentation and test cases, so do not merge)

### CODE REVIEW

**Comment by pmatilai** [comment]
Nice. This is certainly a common need, I've seen activity around this topic on Fedora/RHEL side as well but I've lost track as to where those efforts are now. Tagging @hroncok and @encukou for past interest in this area.

**Comment by encukou** [comment]
Is putting something like `%bcond foo 0%{?default_foo}` in the spec file not an option?

**Comment by hroncok** [comment]
I am confused by the nomenclature here. As a spec author, when I write:

```
%bcond tests 1
```

I expect the tests will run.

Now the distro maintainer (or anybody who can define macros, really) can define a macro:

```
%bcond_default_tests 0
```

What happens to my package? Does it no longer run tests? In my head, this isn't called a *default*, this is an *override*.

-----

A default would be this:

Distribution puts this to their macros:

```
%bcond_default_tests 1
```

A packager can now use:

```
%bcond tests
```

(Notice there is no second argument.) This means "I want this package to run the tests when my distro desires that, but not run them if the distro does not desire that. I explicitly let the distro decision drive my spec behavior. When I care, I use the `--with(out)` command line option explicitly."

That I call a *default*.

(And of course, I realize this would only work for defined defaults, so when the distro does not define a default for tests, it should fail. And if I want to write a spec that anitcipates that possibility, I'd need to write:

```
%bcond tests 0%{?bcond_default_tests}
```

...which I can already do now anyway, as @encukou pointed out.)

**Comment by Vogtinator** [comment]
> Is putting something like `%bcond foo 0%{?default_foo}` in the spec file not an option?

Technically yes, but like bconds itself which are just syntactic sugar, it would be nice to get something like this out of the box and working universally.

Currently the design of bconds makes them effectively worse than just plain macros in my experience. When just using macros like this, it's possible to provide a value in macros, allow overriding that with an `rpmbuild` parameter, specify a default in the .spec file and it can even take arbitrary values instead of just boolean ones!

```
%dnl Provide a default
%{!?enable_foo:%global enable_foo 1}
%dnl Or error out if unset
%{!?enable_foo:%error enable_foo not set}

%if %{enable_foo}
BuildRequires: foo
%else
BuildRequires: bar
%endif
```

The only benefit of bconds is that `rpmbuild` has nicer syntax for `--with(out) foo` and it's easier to translate bconds into commandline switches using the `%{?_with_foo:--foo}` syntax. I'd like something that combines the strengths of simple macros with the special benefits of bconds.

I'm not sure what's easier: Add something to make handling of macros like this easier or extend bconds to allow for more flexibility?

**Comment by mlschroe** [comment]
Sorry for not commenting earlier, this was a busy week.

It's true that this can be done in the specfile, but that can lead to  each individual package maintainer using a different way. I think it's worthwhile that the mechanism is the same for all distributions. The goal is exactly that it works without needing to modify all packages.
(Plus, the proposed pull request also makes it work for bcond_with and bcond_without.)

Regarding nomenclature: I do agree that "default" is not good. I'm open for new suggestions. Here are a couple:
`%bcond_override_libmpeg 1`
`%bcond_force_libmpeg 1`
`%bcond_set_libmpeg 1`

**Comment by hroncok** [comment]
%bcond_set_libmpeg does not carry enough meaning. The other two proposals do and I don't have a preference.

**Comment by pmatilai** [comment]
This is indeed tricky because the desired flow of overrides is different from the usual upstream < distro < host < user override order in rpm. What we have here is something like packager < distro < user - there has to be a way to locally build and override whatever the packager or distro defaults are.

**Comment by ffesti** [comment]
I think I like "override" better. Probably because it emphasizes the out of order operation.

We also need to add this to `manual/conditionalbuilds.md` as soon as we decided on a name.

**Comment by kloczek** [comment]
This PR is wrong and is not solving th issue.
It is the problem that macros used with `%bcond_with foo`/`%bcond_without foo` like `%{?with_foo}` and `%{?without_foo}` are using presence or not that macro.
If you will look closer on macros file
```spec
# Internally, the `--with foo` option defines the macro `_with_foo` and the
# `--without foo` option defines the macro `_without_foo`.
# Based on those and a default (used when neither is given), bcond macros
# define the macro `with_foo`, which should later be checked:

%bcond()        %[ (%2)\
    ? "%{expand:%%{!?_without_%{1}:%%global with_%{1} 1}}"\
    : "%{expand:%%{?_with_%{1}:%%global with_%{1} 1}}"\
]
%bcond_with()           %bcond %{1} 0
%bcond_without()        %bcond %{1} 1

# Shorthands for %{defined with_...}:
%with()         %{expand:%%{?with_%{1}:1}%%{!?with_%{1}:0}}
%without()      %{expand:%%{?with_%{1}:0}%%{!?with_%{1}:1}}
```
you can find that it is used only presence of `%{with_foo}` or not.
This is why I have issue with disable globally %check using %bcond https://github.com/rpm-software-management/rpm/discussions/2870
Please test you PR with test spec and test units which I've added in this discussion and you will find that you PR is not working as expected/

IMO better would be leave %bcond as it is ad define possibility to use something new.

**Comment by ffesti** [comment]
Well turns out this actually works. But may be not the way you expected... and that's why we need docs.

As far as I understand this `%bcond_default` is just a helper macro (and should be prepended with `__`) and the user/distro is supposed to declare %bcond_default_foo manually. With that things work here in my cobbled together test case.

**Comment by ffesti** [comment]
Guess we want `%__bcond_default` and `%bcond_default` for declaring the `%bcond_default_foo` macro even if that is slightly silly.

**Comment by ffesti** [comment]
Meditating a bit more over the name: May be "bcond_default" wasn't as bad as I first thought as the macro does not override the users cli choice but only the packagers default in the spec file. May be this is used seldomly enough to afford "bcond_override_default"?

**Comment by pmatilai** [comment]
I'd like to see tests with real-world bconds in packages. We already have tests for bcond so this can build on top of those. 

I'd like to see documentation and tests on how this is supposed to be used in real life, answering some practical matters like
- Where would the overrides/defaults be stored, a macro file somewhere in the path? 
- Would it be shipped to users so that when you rebuild an src.rpm you get the same selections as the distro build? 
- If I have a bcond set in a spec, overridden to something else by the distro, can I override the distro setting by by a "--with/without foo" on the command line? This seems crucial if the defaults are shipped to users.

Everything is of course simpler if these overrides are only selectively passed by a command line in a central buildsys somewhere, reading up from its own config in whatever format. It makes for an inconsistent rebuild behavior though.

**Comment by ffesti** [comment]
OK, renamed to `bcond_override_default`. This is hopefully used sparingly enough that the additional typing won't kill anyone.

I added some documentation to the Conditional Build page. This should answer the questions above. But someone please prove read them.

**Comment by hroncok** [comment]
Could you please add an example to the docs?

**Comment by ffesti** [comment]
Yes.

File: docs/manual/conditionalbuilds.md [hunk 1]
```diff
@@ -91,5 +91,16 @@ macros which is nicer in other situations, e.g.:
...
+
+For distributions it can be useful to overwrite the build conditionals on a global scale. To not interfere with the users ability to overwrite the conditionals on the command line there is an option to overwrite the default value indenpendent on the one chosen in the spec file.
+
+To do this one can define a `%bcon_override_default_NAME` macro as one or zero or use the `%{bcon_override_default NAME VALUE}` macro. Distributions can put the former into a global macro file that is installed during local builds to propagate these changed defaults outside their build system. Using different versions of the macro file allows building the same set of packages in different ways - e.g. against different libraries - without altering all the spec files.
+
+E.g. add this in the macros file to disable support for zstd assuming this is a common conditional in the distribution:
+```
+%bcon_override_default_zstd 0
+```
+
```
Comment (hroncok) [inline docs/manual/conditionalbuilds.md:104]:
```suggestion

All packages with a `zstd` bcond will now build as if the bcond was defined as `%bcond zstd 0`.
I.e. unless `--with zstd` is used, the bcond will be disabled.

```

File: docs/manual/conditionalbuilds.md [hunk 2]
```diff
@@ -91,5 +91,16 @@ macros which is nicer in other situations, e.g.:
...
 
+## Overrinding Defaults
+
+For distributions it can be useful to overwrite the build conditionals on a global scale. To not interfere with the users ability to overwrite the conditionals on the command line there is an option to overwrite the default value indenpendent on the one chosen in the spec file.
+
+To do this one can define a `%bcon_override_default_NAME` macro as one or zero or use the `%{bcon_override_default NAME VALUE}` macro. Distributions can put the former into a global macro file that is installed during local builds to propagate these changed defaults outside their build system. Using different versions of the macro file allows building the same set of packages in different ways - e.g. against different libraries - without altering all the spec files.
+
+E.g. add this in the macros file to disable support for zstd assuming this is a common conditional in the distribution:
+```
+%bcon_override_default_zstd 0
```
Comment (pmatilai) [inline docs/manual/conditionalbuilds.md:102]: %bcond_, not %bcon_. Multiple cases of that in the above section.

Comment (ffesti) [reply docs/manual/conditionalbuilds.md:102]: Yeah, that looked funny but then I compared it to the ones above...

**Comment by ffesti** [comment]
And here we go. From 2 lines to 120 in just 11 months...

File: tests/rpmbuild.at [hunk 3]
```diff
@@ -2451,6 +2453,105 @@ has_bcond(normally_on)
...
 RPMTEST_CLEANUP
 
+
+
+AT_SETUP([bcond_override_default macros])
+AT_KEYWORDS([bcond build])
+RPMDB_INIT
+
+# check bcond_override_default by defining
+AT_CHECK([
```
Comment (dmnks) [inline tests/rpmbuild.at:2463]: AT_CHECK is deprecated in favor of RPMTEST_CHECK. It's currently just an alias so it works the same but for consistency, we should use the latter.

File: tests/rpmbuild.at [hunk 4]
```diff
@@ -2413,6 +2414,7 @@ runroot rpm -q --provides -p /build/RPMS/noarch/bcondtest-1.0-1.noarch.rpm |
 ],
 [])
 
+
```
Comment (pmatilai) [inline tests/rpmbuild.at:2417]: Couple of unrelated empty lines getting added here and above, and also just before and between the new tests. It's a good idea to eyeball 'git diff' and/or 'git show' output for irregularities like this before pushing.

Other than that, looks fine to me now.
